lib/task_manager.js

```diff
@@ -0,0 +1,50 @@
+class TaskManager {
+  constructor() {
+    this.taskData = {};
+    this.queue;
+    this.buffer = [];
+    this.taskIndex = 0;
+    this.lastPushedToBuffer = 0;
+  }
+
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+    this.queue.on('completed', (data) => this.handleNewData(data));
+  }
+
+  clearQueue() {
+    this.queue.pause();
+    this.queue.clear();
+  }
+
+  retrieveCompleted() {
+    const bufferCopy = [];
+    while (this.buffer.length > 0) bufferCopy.push(this.buffer.shift());
+    return bufferCopy;
+  }
+
+  #pushAllEligable() {
+    while (this.taskData[this.lastPushedToBuffer]) {
+      for (const data of this.taskData[this.lastPushedToBuffer]) this.buffer.push(data);
+      delete this.taskData[this.lastPushedToBuffer];
+      this.lastPushedToBuffer++;
+    }
+  }
+
+  handleNewData([key, newTransformedData]) {
+    this.taskData[key] = newTransformedData;
+    this.#pushAllEligable();
+  }
+
+  pushToQueue(workTask) {
+    this.queue.add(() => {
+      const result = workTask();
+      const currIndex = this.taskIndex;
```

Comment:
This I thought would be enough, but will update this. Seems this is not enough for copying.